### PR TITLE
Use MPI_Bcast instead of multiple p2p messages to update nest from parent

### PR DESCRIPTION
**Pull request by dkokron**
Performance profiling of a HAFS case on NOAA systems revealed significant of time was spent in fill_nested_grid_cpl().  The fill_nested_grid_cpl() routine from FV3/atmos_cubed_sphere/driver/fvGFS/atmosphere.F90 is showing up as a performance bottleneck.  This routine gathers a global SST field (6,336,000 bytes) onto rank 0, then broadcasts that field to all ranks in the nest.  The code uses point-to-point (p2p) messages (Isend/Recv) from rank 0 to the nest ranks.  This communication pattern is maxing out the SlingShot-10 link on the first node resulting in a .15s hit every fifth time step.  This works out to 189 seconds for a full 5-day simulation.  These timings are from the 26 node HAFS case on acorn.

Given the way its currently coded coded, I expect the bottleneck to grow with the number of ranks in the nest.  Timings from a 47-node production case run for 24-hour simulation on cactus with and without using the code changes contained in this PR.

An offline reproducer showed that setting the following environment variables improved the performance of using MPI_Bcast in this scenario.

export MPICH_BCAST_ONLY_TREE=0
export MPICH_COLL_OPT_OFF=mpi_bcast

Scenarios:
1. Unmodified code (Baseline)
2. Unmodified code with above environment settings
3. Code modified following this PR with environment settings
4. Code modified to use MPI_Bcast directly with above environment settings
5. Code modified  to use MPI_Bcast directly without environment settings

Performance metric:
Add up the phase1 and phase2 timings printed in the output listing
grep PASS stdout | awk '{t+=$10;print t}' | tail -1
The units are seconds.

|   Scenario  |   Trial1    |    Trial2   |   Trial3   |   Trial4   |   Trial5   |   Trial6   | Mean | 
| -------------- | ------------ | ----------- |-------------- | ------------ | ----------- |----------- |----------- |
| 1 |  966.2  |  960.9 |
| 2 |  964.9  |  966.9 | 986.6 | 976.4 | x | x | 973.7 |
| 3 |  971.3  | 955.8 |  934.4 | 941.6 | 944.8 | 936.7 | 947.4 |
| 4 |  899.3  | 896.5 | 915.6  | 911.3 | 909.6 | 912.0 | 907.3 |
| 5 |  895.1  | 899.9 |

Based on these timings, I expect savings from scenario 3 (this PR) of ~(26s)*5=131 seconds for a full 5-day simulation.
Based on these timings, I expect savings from scenario 4 (previous version of this PR) of ~(66s)*5=332 seconds for a full 5-day simulation.

This PR does not address any open issues.

I ran the UFS regression suite on acorn and cactus.  Both runs resulted in "REGRESSION TEST WAS SUCCESSFUL"

**Checklist:**

Please check all whether they apply or not
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: model/fv_control.F90
Comment (dkokron) [reply model/fv_control.F90:460]: FMS does have a similar call (mpp_broadcast), but it creates the communicator on-the-fly every time its called.  My testing shows that the creation of the communicator takes about the same amount of time as the actual Bcast.  That's why I create the communicator in the fv_control_init().  Some numbers from an offline reproducer run on acorn.  The offline code creates the MPI communicator every iteration.  The first run takes a long time (cycles) probably due to MPI initialization.  "setup subcomm" refers to the call to MPI_Comm_split()

Cycles to setup subcomm: 845895398
Cycles Bcast with a comm: 352677465

Subsequent calls within the same execution settle into a nice pattern.
Cycles to setup subcomm: 35054707
Cycles Bcast with a comm: 3219772

Cycles to setup subcomm: 1966522
Cycles Bcast with a comm: 3727102

Cycles to setup subcomm: 2671380
Cycles Bcast with a comm: 3211785

Cycles to setup subcomm: 1973183
Cycles Bcast with a comm: 3195247

Cycles to setup subcomm: 1884240
Cycles Bcast with a comm: 3205283

Cycles to setup subcomm: 2135385
Cycles Bcast with a comm: 3193492

Cycles to setup subcomm: 2105010
Cycles Bcast with a comm: 3203685

Comment (bensonr) [reply model/fv_control.F90:460]: @dkokron - please send my your test program so I can confirm your findings with mpp_broadcast.  As I look at the code, if the pelist has been defined previously, it should be using a predefined communicator.  A new communicator is defined only when a new, unique pelist is encountered.

Comment (dkokron) [reply model/fv_control.F90:460]: It's on cactus.  I will upload when cactus comes back from dedicated time.

Perhaps then it's enough to create a pelist with the appropriate PEs in fv_control_init() instead of an MPI_Comm.

Comment (bensonr) [reply model/fv_control.F90:460]: Thanks for letting me know about the cactus downtime.

There should be no need to create the pelist in fv_control.  It can be defined/used within the fill_nested_grid_cpl function.  The call to get_peset [here](https://github.com/NOAA-GFDL/FMS/blob/main/mpp/include/mpp_transmit_mpi.fh#L177) will execute this code in [mpp_util_mpi.inc](https://github.com/NOAA-GFDL/FMS/blob/main/mpp/include/mpp_util_mpi.inc#L91-L139).  If the pelist is recognized, it will return the mpp-managed internal list location and mpp_broadcast will use the communicator associated with that list entry.  If the pelist isn't recognized, it will add it to the linked list of known pelists and create a communicator via MPI_Comm_create_group.  If you want to look over the code with me, let me know and we can schedule it.

Comment (dkokron) [reply model/fv_control.F90:460]: I see your point about being able to use get_peset() to create the communicator once storing it in the pset structure, then mpp_broadcast() reuses that communicator.  My implementation does essentially the same thing while storing the communicator in the fv_atmos_type structure.

I'm a little worried about the extra work done in the MPP approach.  Every call to get_peset() will check for monotonicity and 
check for a match to an existing peset.  I'll see if I can code up the MPP approach in my unit tester and determine if my worry is justified.

Comment (dkokron) [reply model/fv_control.F90:460]: My worry was about the extra work was unfounded.  I will re-code the PR using mpp_broadcast.

You can find my unit testers under clogin02:/u/daniel.kokron/Bcast/
The testers simulate what happens in a HAFS case with one nest run on 736 ranks.

Some timings:
grep Cycles Bcast.o58131230

Using MPI_Bcast with a pre-made communicator (C implementation)
Cycles to setup subcomm: 789298313
 Cycles Bcast with a comm: 350650485
 Cycles Bcast with a comm: 96031373
 Cycles Bcast with a comm: 3138885
 Cycles Bcast with a comm: 3160485
 Cycles Bcast with a comm: 3150698
 Cycles Bcast with a comm: 3127387
 Cycles Bcast with a comm: 3128850
 Cycles Bcast with a comm: 3114293
 Cycles Bcast with a comm: 3123540
 Cycles Bcast with a comm: 3147165
 Cycles Bcast with a comm: 3136793
 Cycles Bcast with a comm: 3124845
 Cycles Bcast with a comm: 3137737
 Cycles Bcast with a comm: 3132450
 Cycles Bcast with a comm: 3156840

Using MPI_Bcast with a pre-made communicator (Fortran implementation)
 Cycles to setup subcomm    136254240
 Cycles rank/thread    332684775
 Cycles rank/thread      4624537
 Cycles rank/thread      3276248
 Cycles rank/thread      3257415
 Cycles rank/thread      3283357
 Cycles rank/thread      3271545
 Cycles rank/thread      3424860
 Cycles rank/thread      3280073
 Cycles rank/thread      3279398
 Cycles rank/thread      3288915
 Cycles rank/thread      3359723
 Cycles rank/thread      3296475
 Cycles rank/thread      3246457
 Cycles rank/thread      3312000
 Cycles rank/thread      3278633

Using mpp_broadcast() 
Cycles rank/thread    548630797
 Cycles rank/thread      3106642
 Cycles rank/thread      3273503
 Cycles rank/thread      3308108
 Cycles rank/thread      3264412
 Cycles rank/thread      3284077
 Cycles rank/thread      3473212
 Cycles rank/thread      3260768
 Cycles rank/thread      3294653
 Cycles rank/thread      3311325
 Cycles rank/thread      3297983
 Cycles rank/thread      3296722
 Cycles rank/thread      3306937
 Cycles rank/thread      3325117
 Cycles rank/thread      3270375

Comment (bensonr) [reply model/fv_control.F90:460]: @dkokron - I appreciate your willingness to work with us on this.

Comment (dkokron) [reply model/fv_control.F90:460]: Never easy.

Swapping in mpp_broadcast in place of MPI_Bcast in fill_nested_grid_cpl() results in runtime failure.

```
   subroutine fill_nested_grid_cpl(this_grid, proc_in)
-    use mpi
+    !use mpi
+    use mpp_mod, only: mpp_broadcast
     integer, intent(in) :: this_grid
     logical, intent(in), optional :: proc_in
 
@@ -2460,7 +2461,8 @@ contains
                             g_dat(isg:,jsg:,1), position=CENTER)
     endif
     if(any(mpp_pe() == Atm(this_grid)%Bcast_ranks)) then
-      call MPI_Bcast(g_dat, size(g_dat), MPI_REAL, 0, Atm(this_grid)%Bcast_comm, ierr) ! root==0 because sending rank (Atm(this_grid)%sending) is rank zero in Bcast_comm
+      !call MPI_Bcast(g_dat, size(g_dat), MPI_REAL, 0, Atm(this_grid)%Bcast_comm, ierr) ! root==0 because sending rank (Atm(this_grid)%sending) is rank zero in Bcast_comm
+      call mpp_broadcast(g_dat, size(g_dat),Atm(this_grid)%sending_proc,Atm(this_grid)%Bcast_ranks)
     endif
     call timing_off('COMM_TOTAL')
     if (process) then
```
The stack trace points to a part of the code I didn't change.

> nid001001.acorn.wcoss2.ncep.noaa.gov 0: NOTE from PE     0:  eval_move_nest: move_cd_x=           0 move_cd_y=           0 do_move= F
> nid001001.acorn.wcoss2.ncep.noaa.gov 0: NOTE from PE     0: Opened BC file: INPUT/gfs_bndy.tile7.003.nc
> nid001001.acorn.wcoss2.ncep.noaa.gov 0: forrtl: error (78): process killed (SIGTERM)
> nid001001.acorn.wcoss2.ncep.noaa.gov 0: Image              PC                Routine            Line        Source             
> nid001001.acorn.wcoss2.ncep.noaa.gov 0: fv3_32bit.exe      000000000258B397  fv_regional_mod_m        3387  fv_regional_bc.F90
> nid001001.acorn.wcoss2.ncep.noaa.gov 0: fv3_32bit.exe      000000000257BCD5  fv_regional_mod_m        1792  fv_regional_bc.F90
> nid001001.acorn.wcoss2.ncep.noaa.gov 0: fv3_32bit.exe      000000000259E347  fv_regional_mod_m        1592  fv_regional_bc.F90
> nid001001.acorn.wcoss2.ncep.noaa.gov 0: fv3_32bit.exe      000000000291AABA  atmosphere_mod_mp         673  atmosphere.F90

**Comment by lharris4** [comment]
Hi. Sorry for the delay. This is a bit odd and might be a sign of a memory
bug somewhere. How soon in the run does this error trigger? Also, does
running in debug mode raise any errors?

Thanks,
Lucas

**Comment by dkokron** [comment]
I was able to get it working with mpp_broadcast() by playing games with mpp_set_current_pelist().  I'm running the regression suite now and will redo some of the performance testing to make sure it still helps.

**Comment by lharris4** [comment]
That is great. Thanks for continuing to work on this.

Lucas

**Comment by dkokron** [comment]
I gathered more timings and updated the table.  Using MPI_Bcast directly is significantly faster than using mpp_broadcast().  Please review the code and let me know if I'm using MPP incorrectly.

**Comment by bensonr** [comment]
@dkokron - what timings did you get using mpp_broadcast for comparison with the data in the PR description

**Comment by dkokron** [comment]
Scenario 3

**Comment by dkokron** [comment]
Given the significant difference in timings between scenarios 3 (mpp_braodcast) and 4 (MPI_Bcast), which proposal should we move froward with?

**Comment by lharris4** [comment]
Hello @dkokron. @bensonr and I are discussing how to move forward. There may be an underlying problem in the way the SST is being gathered in the first place that isn't addressed by this fix.

**Comment by dkokron** [comment]
What do you have in mind?
Let me know if you need me to test anything.

**Comment by bensonr** [comment]
@dkokron - I have a test case that creates a 6 tile mosaic with a nest.
I'd like to run your exact sizes to get some timings on gaea.  What core
counts and nest configuration are you using so I can have a good comparison.

**Comment by bensonr** [comment]
I gave up my access to NCEP/NCO when WCOSS2 came online and the rules and
restrictions became too much for me to maintain it.  I'd appreciate if you
could give me the domain sizes (npx,npy,npz) and layouts for the global and
nest, as well as the nest_nml info for the two cases you are interested in
- thx


The 26-node case is on acorn

**Comment by dkokron** [comment]
26-node case
Parent grid
  layout = 16,28  ! LongEW
  io_layout = 1,1
  npx = 1321
  npy = 1201
  ntiles = 1
  npz = 81

  grid_pes = 448,288

Nest grid
  layout = 12,24 ! LongEW
  io_layout = 1,1
  npx = 601
  npy = 601
  ntiles = 1
  npz = 81


47-node case
------------------
Parent grid
  layout = 36,50
  io_layout = 1,1
  npx = 1201
  npy = 1201
  ntiles = 1
  npz = 81

  grid_pes = 1800,840

Nest grid
  layout = 20,42
  io_layout = 1,1
  npx = 601
  npy = 601
  ntiles = 1
  npz = 81

**Comment by bensonr** [comment]
Thanks for passing all of this along.  I finally have the prototype test
code running where I create a top-level domain with a nest.  Using your
data for the 47-node case (21 nodes on gaea with no OMP threads), I have
generated data for both 32-bit and 64-bit quantities.  The sample sizes
were 25 and 50 respectively.  The first iteration for each was a touch
longer as they set up the pelist and communicator.

For 64-bit data, the max average time for:
   global gather on 1800 parent ranks is:  0.16s
   broadcast from 1 parent rank to 840 nest ranks is:  0.058s

For 32bit data, the max average time for:
   global gather on 1800 parent ranks is: 0.109s
   broadcast from 1 parent rank to 840 nest ranks is:  0.029s

In your code you are swapping scope via the mpp_set_current_pelist call.
There is no need for the scope changes when using explicit pelists for mpp
communication.

          do n=1,50

            t_dum = mpp_get_tile_id(domain_coarse)

            if (((t_dum(1) == tile_coarse(1)) .and.
mpp_domain_is_tile_root_pe(domain_coarse)) .or. (is_fine_pe)) then

              time_it = MPI_Wtime()

              bcast_npes = mpp_get_domain_npes(domain_fine) + 1

              allocate(bcast_pelist(bcast_npes))

              bcast_pelist(1) = mpp_get_domain_tile_root_pe(domain_coarse)

              call mpp_get_pelist(domain_fine, bcast_pelist(2:bcast_npes))

              call mpp_broadcast(xg, size(xg), bcast_pelist(1),
bcast_pelist)

              time_it = MPI_Wtime() - time_it

              t_max=time_it

              t_min=time_it

              call mpp_max(t_max,bcast_pelist)

              call mpp_min(t_min,bcast_pelist)

              if (mpp_pe()==bcast_pelist(1)) write(6,'(a,f10.6,2x,f10.6)')
' time for bcast ', t_min, t_max

              deallocate(bcast_pelist)

            endif

          enddo

In the code above, all MPI-ranks enter the do-loop but only those
participating in the broadcast enter the if-test.  The above code is not
100% generic, though you can simplify it immensely for use in
fill_nested_grid_cpl

**Comment by dkokron** [comment]
Rusty,
I was getting runtime errors without the scoping.  I think I see my error.
I am starting my bcast_npes indexing at zero not 1.  More testing.
Dan

**Comment by dkokron** [comment]
Rusty,
Please send me your complete reproducer.

**Comment by bensonr** [comment]
Sure, I'll share it via your email.

**Comment by dkokron** [comment]
***@***.*** or
***@***.***

**Comment by bensonr** [comment]
Unfortunately, github blocked the addresses you just included.  I was planning to send to your NOAA account, if that's okay with you.

**Comment by dkokron** [comment]
oops ***@***.*** (not gov)
NOAA account works too.

**Comment by dkokron** [comment]
This is what I get when using mpp_broadcast() in HAFS without the scoping.
I really cannot imagine how that -448 is getting in there.

nid001022.acorn.wcoss2.ncep.noaa.gov 724:  fill_nested_grid_cpl:
 724
nid001022.acorn.wcoss2.ncep.noaa.gov 724: MPICH ERROR [Rank 724] [job id
65bd0239-958a-4cad-bd26-b2808f73b8ba] [Fri Jun  2 19:53:29 2023]
[nid001022] - Abort(940145670) (rank 724 in comm 0): Fatal error in
PMPI_Group_incl: Invalid rank, error stack:
nid001022.acorn.wcoss2.ncep.noaa.gov 724:
PMPI_Group_incl(170).............: MPI_Group_incl(group=0xc8000006, n=289,
ranks=0x7fffc9caf880, new_group=0x14713ad803f8) failed
nid001022.acorn.wcoss2.ncep.noaa.gov 724:
MPIR_Group_check_valid_ranks(244): Invalid rank in rank array at index 0;
value is -448 but must be in the range 0 to 287
nid001022.acorn.wcoss2.ncep.noaa.gov 724:
nid001022.acorn.wcoss2.ncep.noaa.gov 724: aborting job:

    if(any(mpp_pe() == Atm(this_grid)%Bcast_ranks)) then
      print*,'fill_nested_grid_cpl: ',mpp_pe()
      call mpp_broadcast(g_dat, size(g_dat),Atm(this_grid)%Bcast_ranks(1),
Atm(this_grid)%Bcast_ranks)
    endif

**Comment by dkokron** [comment]
Submitted https://github.com/NOAA-GFDL/FMS/pull/1246 to address the failure seen above when using mpp_broadcast().

Please consider using the pure MPI approach to address this performance issue.
We can revisit mpp if that PR is accepted, released, and approved by NOAA.

**Comment by dkokron** [comment]
This PR now depends on https://github.com/NOAA-GFDL/FMS/pull/1246.  My understanding is that UFS/HAFS will need to wait until FMS releases a new tag.  That tag will then need go through the NOAA approval and installation process.

**Comment by thomas-robinson** [comment]
@dkokron I mentioned you in the https://github.com/NOAA-GFDL/FMS/pull/1246 earlier.  We are willing to work with your timeline in FMS and issue a patch for you instead of waiting for the full release if you need it now.

**Comment by dkokron** [comment]
Tom,
Many thanks for being willing to push out a patch.
Since this is targeted at the HAFS application, I'm hoping to get some
feedback from them regarding the urgency of this.  I will report back if I
hear anything from the HAFS team.
Dan

**Comment by lharris4** [comment]
Hi, Dan, Tom. Thanks for working on this.

Lucas

**Review by lharris4 — APPROVED** [review]
I think this should be a good solution now. Thanks for working on this.

**Comment by laurenchilutti** [comment]
@dkokron We will never merge a pull request targeting the dev/emc branch without specific instruction to do so from the code manager of the UFS weather model.

**Comment by laurenchilutti** [comment]
@dkokron I also commented this on the FMS PR, but please let us know when you hear back from HAFS regarding the timeline in which you need this change (which will require https://github.com/NOAA-GFDL/FMS/pull/1246).  If you do require it earlier then our next planned release of FMS (which is approx. mid July) we would then need about 1 week to issue a patch for you.

**Comment by dkokron** [comment]
> @dkokron We will never merge a pull request targeting the dev/emc branch without specific instruction to do so from the code manager of the UFS weather model.

@laurenchilutti Did I do something wrong in my fork that caused you to say this?

**Comment by laurenchilutti** [comment]
> > @dkokron We will never merge a pull request targeting the dev/emc branch without specific instruction to do so from the code manager of the UFS weather model.
> 
> @laurenchilutti Did I do something wrong in my fork that caused you to say this?

No! I just saw that you wrote " Please hold off on doing anything with this PR" and I wanted to reassure you that we won't merge it until you are ready.

**Comment by dkokron** [comment]
Cool.

**Comment by dkokron** [comment]
I noticed that FMS-2023.02 is released and includes the fix for https://github.com/NOAA-GFDL/FMS/pull/1246.  The trouble is that they also included a breaking change that prevents the UFS code from building (at least not without building FMS with with the indicated flag).  Specifically, the following test in the release nodes.

FV3/atmos_cubed_sphere/model/multi_gases.F90(45): error #6580: Name in only-list does not exist or is not accessible.   [OPEN_NAMELIST_FILE]
      use fms_mod,           only: check_nml_error, open_namelist_file, close_file
----------------------------------------------------^
/u/daniel.kokron/Scratch/Projects/UFS/hafs-BcastFillNestedGridCpl/FV3/atmos_cubed_sphere/model/multi_gases.F90(45): error #6580: Name in only-list does not exist or is not accessible.   [CLOSE_FILE]
      use fms_mod,           only: check_nml_error, open_namelist_file, close_file


FMS_IO/MPP_IO: The two older io modules, fms_io_mod and mpp_io_mod, have been deprecated and will not be compiled by default. If you wish to compile these modules, you must use the -Duse_deprecated_io CPP flag or the --enable-deprecated-io configure option if building with autotools.

How do I  move forward with this PR?  Should I wait until UFS is ported to the new FMS or should I just rebuild FMS to include the deprecated features?

**Comment by laurenchilutti** [comment]
@dkokron This latest FMS update removed the old fmsio with the newer fms2io.  About a year or more ago we had updated the FV3 code to use the latest fms2io code, clearly there are more updates needed.  I can give this a look  and try to see if I can easily fix these compilation errors.  I will also bring this up to the UFS code management meeting which is this Friday.

**Comment by bensonr** [comment]
@dkokron - the multi_gases file needs to be updated to remove the option for all cores to open/read/close the namelist files .  The preferred method for reading namelist files is to use the internal file read via the _input_nml_file_ character array.  To do this, you need to remove the #else block from the #ifdef/#endif pair in [multi_gases.F90](https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/blob/dev/emc/model/multi_gases.F90#L157-#L171) as well as the supporting fortran module includes.

**Comment by laurenchilutti** [comment]
I think I was able to make all of the necessary changes for FV3, but GAEA is down today and I cannot test.  @dkokron if you have access to a different UFS supported machine, you could try to merge my changes with yours and test.  https://github.com/laurenchilutti/GFDL_atmos_cubed_sphere/tree/deprecated_io_emc

@bensonr There is still a possibility that other UFS components are using deprecated FMS subroutines/functions.

**Comment by dkokron** [comment]
I've worked though compile issues using https://github.com/laurenchilutti/GFDL_atmos_cubed_sphere/tree/deprecated_io_emc as a guide.  I have pushed my changes at and below the atmos_cubed_sphere level to my branch.
Currently working through issues at the fv3atm level.

**Comment by dkokron** [comment]
I finished porting FV3 to use FMS-2023.02.  The main test case I've been using runs fine and performs as expected.  What other testing is needed to move forward with this PR?

I also ran the UFS regression suite using my branch associated with this PR.  There are numerous compile failures due to other components (MOM6, NOAHMP and CMEPS) needing to be ported to FMS-2023.02 (see attached).  I'll take those issues up with those developers.
[NeedsPort.txt](https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/files/12339987/NeedsPort.txt)

**Comment by junwang-noaa** [comment]
@dkokron Thanks for running the UFS WM regression test. 
@jiandewang @DeniseWorthen @uturuncoglu @barlage would you please take a look at the NeedsPort.txt list? This will help us to move to fms-2023.02 in UFS WM. Thanks

**Comment by bensonr** [comment]
@junwang-noaa - I know this is a little off topic for the dycore repo here...  While you are working through that list, you still have the ability to compile in the deprecated io modules via a -Duse_deprecated_io flag (see [FMS release notes](https://github.com/NOAA-GFDL/FMS/releases/tag/2023.02)).  We test MOM6 as part of our FMS release testing and there are some updates from the main trunk that will need to be pulled into the ufs branch to resolve any issues.

**Comment by dkokron** [comment]
As far as GFDL_atmos_cubed_sphere is concerned, is this PR fit to move forward?

**Comment by laurenchilutti** [comment]
As long as FMS is compiled with the flag -Duse_deprecated_io, this PR is fit to move forward.  @jkbk2004 as UFS code manager, when do you think we can have this PR merged?

**Comment by jkbk2004** [comment]
We are working on Spack stack PR. For FMS-2023-02, JEDI team need to use -Duse_deprecated_io. A bit coordination and test is needed to confirm the test results. I think we can schedule to commit next week or so. But I will keep checking as test going on.

**Comment by laurenchilutti** [comment]
> In that case, should I extract the FMS-2023.02 porting modifications from my branch?

I see no harm in keeping these changes in.

File: model/fv_control.F90
Comment (bensonr) [inline model/fv_control.F90:745]: I think this logic could be simplified by using the entries in the Atm type.  Something like:

if (Atm(n)%neststruct%parent_tile == mpp_get_tile_id(Atm(n)%parent_grid%domain)) Atm(n)%Bcast_ranks(1) = mpp_get_domain_tile_root_pe(Atm(n)%parent_grid%domain)

File: tools/fv_grid_tools.F90
Comment (laurenchilutti) [reply tools/fv_grid_tools.F90:234]: This is due to fmsio and old mosaic_mod being deprecated.  This is the same approach done in the main and dev/gfdl branches

Comment (bensonr) [reply tools/fv_grid_tools.F90:234]: @laurenchilutti - thanks I should've checked there first.

Comment (dkokron) [reply tools/fv_grid_tools.F90:234]: That was part of my FMS port.  I offered to yank those mods, but Lauren said not to bother.  I have no preference.

File: model/fv_arrays.F90
Comment (dkokron) [reply model/fv_arrays.F90:1851]: Not needed anymore since we are using mpp_broadcast() now instead of MPI_Bcast().

File: model/fv_control.F90
Comment (dkokron) [reply model/fv_control.F90:745]: I think I yanked that logic from atmos_cubed_sphere/tools/fv_restart.F90 line 885.  I'll try a compile/run using your suggestion.

Comment (dkokron) [reply model/fv_control.F90:745]: atmos_cubed_sphere/model/fv_control.F90(742): error #6512: A scalar-valued expression is required in this context.
`if(Atm(n)%neststruct%parent_tile == mpp_get_tile_id(Atm(n)%parent_grid%domain)) then`
-------------------------------------^
I'll look at it more tomorrow.

Comment (dkokron) [reply model/fv_control.F90:745]: mpp_get_tile_id() returns an integer array.
line 1113 of mpp/include/mpp_domains_util.inc
`integer, dimension(size(domain%tile_id(:))) :: mpp_get_tile_id`
so I need to dump the output to the existing tile_id(1) variable.

my proposal becomes
`tile_id = mpp_get_tile_id(Atm(n)%domain)`
`if( Atm(n)%neststruct%parent_tile == tile_id(1) ) then`
`  Atm(n)%Bcast_ranks(1) = mpp_get_domain_tile_root_pe(Atm(n)%parent_grid%domain)`
`endif`

Comment (bensonr) [reply model/fv_control.F90:745]: Thanks for working through the issues with my napkin programming.  If that works and looks cleaner to you, then I'll approve the PR

Comment (dkokron) [reply model/fv_control.F90:745]: At least for the HAFS case I have, the above 'if' test is failing and results in the Bcast_ranks(1) entry not getting set.  My current code looks like this. 

`        ! parent grid sending rank within Bcast_ranks array`
`        tile_id = mpp_get_tile_id(Atm(n)%domain)`
`        if( Atm(n)%neststruct%parent_tile == tile_id(1) ) then`
`          write(6,'("fv_control_init: Setting lead rank for grid using new way ",2I4)') n,Atm(n)%neststruct%parent_tile`
`           Atm(n)%Bcast_ranks(1) = mpp_get_domain_tile_root_pe(Atm(n)%parent_grid%domain)`
`          else`
`           write(6,'("fv_control_init: Setting lead rank for grid old way ",2I4)') n,Atm(n)%neststruct%parent_tile`
`          Atm(n)%Bcast_ranks(1)=Atm(n)%parent_grid%pelist(1) + &`
`                   ( Atm(n)%neststruct%parent_tile-tile_fine(Atm(n)%parent_grid%grid_number)+ &`
`                      Atm(n)%parent_grid%flagstruct%ntiles-1 )*Atm(n)%parent_grid%npes_per_tile`
`          endif`

 I don't feel like debugging the new approach, so I vote keeping the original (ugly) approach.

Comment (bensonr) [reply model/fv_control.F90:745]: I think I see the issue with your modification, but don't think it's worth continuing to debug.  For documentation purposes, I think the if-test should be an ANY and not specifically against tile_id(1).

Comment (dkokron) [reply model/fv_control.F90:745]: Will test with any()

Comment (dkokron) [reply model/fv_control.F90:745]: Still going through the ugly path.

`if( any(Atm(n)%neststruct%parent_tile == tile_id) ) then`

Comment (bensonr) [reply model/fv_control.F90:745]: I think the tile_id should for the parent grid, i.e. _mpp_get_tile_id(Atm(n)%parent_grid%domain)_.  My example with the mismatched if-test was using that domain.

Comment (dkokron) [reply model/fv_control.F90:745]: `tile_id = mpp_get_tile_id(Atm(n)%parent_grid%domain)`

Does run, however, the parent grid ranks are still going through the old path while the nest ranks go through the new path.

Comment (dkokron) [reply model/fv_control.F90:745]: Can we move forward with the ugly code?

File: model/fv_arrays.F90
Comment (bensonr) [reply model/fv_arrays.F90:1851]: Would like to see this _use_ statement removed prior to accepting.

Comment (dkokron) [reply model/fv_arrays.F90]: Removed unneeded _use_ of MPI module in model/fv_arrays.F90

**Comment by laurenchilutti** [comment]
@bensonr are there anymore changes that you would like to see in this PR?

**Review by bensonr — APPROVED** [review]

**Comment by dkokron** [comment]
The FMS modification that supports this PR was merged.  See details at https://github.com/NOAA-GFDL/FMS/pull/1246

**Comment by bensonr** [comment]
@dkokron - at the bi-weekly UFS code managers meeting we tell them this is ready to merge - they have not yet blessed it.

**Comment by laurenchilutti** [comment]
The UFS code managers will be ready to merge this change in in the coming weeks.  @dkokron could you create a fv3atm PR and UFS-weather-model PR for this change?

**Comment by dkokron** [comment]
@laurenchilutti Will do.

**Comment by dkokron** [comment]
Opened PR against fv3atm
https://github.com/NOAA-EMC/fv3atm/pull/743

**Comment by dkokron** [comment]
Opened PR with UFS
https://github.com/ufs-community/ufs-weather-model/pull/2059

**Comment by FernandoAndrade-NOAA** [comment]
Testing on #[2059](https://github.com/ufs-community/ufs-weather-model/pull/2059) has completed successfully, please proceed with the merge process for this PR.

**Comment by zach1221** [comment]
> Testing on #[2059](https://github.com/ufs-community/ufs-weather-model/pull/2059) has completed successfully, please proceed with the merge process for this PR.

Copying in @laurenchilutti as well for this request.

**Comment by jkbk2004** [comment]
@bensonr can you merge this pr?

**Pull request merged**
